Add `emitDispEmbBroadcastCount` to display the embedded broadcasted element count.

Ruihan-Yin opened the pull request:
# Description 

JitDisasm doesn't include embedded broadcast syntax in disassembly.
 Eg.
`vaddps   zmm0, zmm0, dword ptr [reloc @RWD00]`

This PR will add the missing tag with the broadcasted element count, so the new disassembly will look like:

`vaddps   zmm0, zmm0, dword ptr [reloc @RWD00] {1to16}`

Ruihan-Yin committed:
Add `emitDispEmbBroadcastCount` to display
the embedded broadcasted element count.

BruceForstall commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -10722,6 +10722,20 @@ void emitter::emitDispInsHex(instrDesc* id, BYTE* code, size_t sz)
+void emitter::emitDispEmbBroadcastCount(instrDesc* id)
+{
+    ssize_t baseSize = GetInputSizeInBytes(id);
+    id->idClearEvexbContext();
+    ssize_t vectorSize = (ssize_t)emitGetMemOpSize(id);
+    id->idSetEvexbContext();
```
It seems like we shouldn't be altering the instrDesc during a debug-only print function. Ideally, this function would take a `const instrDesc*`.

Can you refactor `emitGetMemOpSize()` to be:

```
if (broadcast)
{
   return <broadcast result>
}
else
{
   return emitGetBaseMemOpSize(); // all the non-broadcast logic
}
```

Then, here:


```suggestion
    ssize_t vectorSize = (ssize_t)emitGetBaseMemOpSize(id);
```

Also, could this also do the Evexb check:

```
if (!id->idIsEvexbContext())
  return;

// otherwise, do the Evexb print logic
```

so the callers don't need to duplicate all the Evexb checking logic?

tannergooding commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -10722,6 +10722,20 @@ void emitter::emitDispInsHex(instrDesc* id, BYTE* code, size_t sz)
+    id->idClearEvexbContext();
+    ssize_t vectorSize = (ssize_t)emitGetMemOpSize(id);
+    id->idSetEvexbContext();
+    printf(" {1to%d}", vectorSize / baseSize);
```
Can we make this a comment instead, so that it stays mostly valid assembly?

Something like `; broadcast x%d` instead, so this would be `vaddps   zmm0, zmm0, dword ptr [reloc @RWD00] ; broadcast x16`

BruceForstall replied:
Isn't the `{1toX}` notation itself valid assembly?

tannergooding replied:
Ah, so it is. I've seen compilers not include that and instead infer it from the address size (e.g. `dword ptr [addr]` vs `zmmword ptr [addr]`).

Ruihan-Yin committed:
Resolve comments:
rename the previous `emitGetMemOpSize` to `emitGetBaseMemOpSize`,
the new `emitGetMemOpSize` will check if EVEX.b is set and return the
memory size accordingly.

Ruihan-Yin committed:
bug fixes:
added 2 missing break

BruceForstall commented on src/coreclr/jit/emit.h:
```diff
@@ -1541,6 +1541,12 @@ class emitter
             _idEvexbContext = 1;
             assert(_idEvexbContext == 1);
         }
+        void idClearEvexbContext()
+        {
+            assert(_idEvexbContext == 1);
+            _idEvexbContext = 0;
+            assert(_idEvexbContext == 0);
+        }
```
This is no longer needed

BruceForstall commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -11123,6 +11136,11 @@ void emitter::emitDispIns(
         {
             printf("%s, %s, %s", emitRegName(id->idReg1(), attr), emitRegName(id->idReg2(), attr), sstr);
             emitDispAddrMode(id);
+            if (!id->idIsEvexbContext())
+            {
+                break;
+            }
+            emitDispEmbBroadcastCount(id);
```
I think I was unclear in my previous suggestion. I was suggesting that here, in emitDispIns, we just have:
```suggestion
            emitDispEmbBroadcastCount(id);
```

And add:
```
            if (!id->idIsEvexbContext())
            {
                return;
            }
```
at the very beginning of `emitDispEmbBroadcastCount()`.

Ruihan-Yin replied:
Ah, yes, this is more clear. I understood it wrong.

Ruihan-Yin committed:
resolve comments:
1. remove un-needed helper function.
2. move the Evexb check into the display function.
